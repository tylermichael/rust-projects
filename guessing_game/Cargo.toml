[package]
name = "guessing_game"
version = "0.1.0"
edition = "2021"

[profile.dev.package]
insta.opt-level = 3
similar.opt-level = 3

[dependencies]
rand = "0.8.5"

[dev-dependencies]
insta = "1.34.0"
