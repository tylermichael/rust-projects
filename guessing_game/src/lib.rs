use std::{
    cmp::Ordering,
    io::{BufRead, Write},
};

pub fn do_guess_random<R, W>(mut reader: R, mut writer: W, num: i32) -> Result<(), std::io::Error>
where
    R: BufRead,
    W: Write,
{
    writeln!(&mut writer, "The random number is: {num}")?;
    writeln!(&mut writer, "Enter a number from 1-100: ")?;
    loop {
        let mut guess = String::new();
        reader.read_line(&mut guess).expect("failed to read stdin");
        let guess: i32 = match guess.trim().parse() {
            Ok(n) => n,
            Err(_) => {
                writeln!(&mut writer, "You must enter a number.")?;
                continue;
            }
        };

        match num.cmp(&guess) {
            Ordering::Equal => {
                writeln!(&mut writer, "You guessed correctly!")?;
                return Ok(());
            }
            Ordering::Greater => writeln!(&mut writer, "The number is greater")?,
            Ordering::Less => writeln!(&mut writer, "The number is lower")?,
        }
    }
}

#[test]
fn test_in_memory() {
    let guess = b"10";
    let mut output = Vec::new();

    let res = do_guess_random(&guess[..], &mut output, 10);

    assert!(res.is_ok());
    insta::assert_debug_snapshot!(String::from_utf8(output));
}

#[test]
fn test_higher_lower() {
    let guess = b"9\n11\n10";
    let mut output = Vec::new();

    let res = do_guess_random(&guess[..], &mut output, 10);

    assert!(res.is_ok());
    insta::assert_debug_snapshot!(String::from_utf8(output));
}

#[test]
fn test_bad_input() {
    let guess = b"not a number\n10";
    let mut output = Vec::new();

    let res = do_guess_random(&guess[..], &mut output, 10);

    assert!(res.is_ok());
    insta::assert_debug_snapshot!(String::from_utf8(output));
}
