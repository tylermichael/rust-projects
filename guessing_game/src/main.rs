use guessing_game::*;
use rand::Rng;
use std::io::{stdin, stdout};

fn main() -> Result<(), std::io::Error> {
    let num: i32 = rand::thread_rng().gen_range(1..=100);
    do_guess_random(stdin().lock(), stdout().lock(), num)
}
