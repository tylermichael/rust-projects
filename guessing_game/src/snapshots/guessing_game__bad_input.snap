---
source: src/main.rs
expression: "String::from_utf8(output)"
---
Ok(
    "The random number is: 10\nEnter a number from 1-100: \nYou must enter a number.\nYou guessed correctly!\n",
)
